src/dashboard/src/components/archival_storage/atom.py

```diff
@@ -0,0 +1,104 @@
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Archivematica is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Archivematica.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import os
+import tempfile
+
+from agentarchives.atom.client import AtomClient, AtomError
+from metsrw import METSDocument, MetsError
+from storageService import extract_file
+from slumber.exceptions import SlumberHttpBaseException
+
+from components import helpers
+from main.models import DashboardSetting
+
+
+logger = logging.getLogger('archivematica.dashboard')
+
+
+def get_atom_client():
+    url = helpers.get_setting('dip_upload_atom_url')
+    key = helpers.get_setting('dip_upload_atom_key')
+    return AtomClient(url, key)
+
+
+class AtomMetadataUploadError(Exception):
+    pass
+
+
+def upload_dip_metadata_to_atom(aip_name, aip_uuid, parent_slug):
+    """
+    Write to a AtoM's resource (parent_slug) the metadata of the objects of a
+    AIP given its name and UUID. Return the slug of the new container resource
+    created to hold the metadata objects.
+    """
+    try:
+        with tempfile.NamedTemporaryFile() as temp:
+            # Download METS file
+            mets_path = '{}-{}/data/METS.{}.xml'.format(aip_name, aip_uuid, aip_uuid)
+            logger.debug('Extracting file %s into %s', mets_path, temp.name)
+            extract_file(aip_uuid, mets_path, temp.name)
+
+            client = get_atom_client()
+            mw = METSDocument.fromfile(temp.name)
+
+            # Create file container
+            logger.info('Creating file container with slug %s and title %s', parent_slug, aip_name)
+            file_slug = client.add_child(parent_slug=parent_slug, title=aip_name, level='File')
+
+            def add_prop_from_xml(dict_, name, el, xpath):
+                """
+                Write to a dictionary a new pair with the given key and the value
+                taken from the text attribute of the element matched by the given
+                XPath query.
+                """
+                res = el.find(xpath)
+                if res is not None and res.text:
+                    dict_[name] = res.text
+                    logger.debug('Extracted property %s from METS: %s', name, res.text)
+                logger.debug('Failed to extract property %s from METS: not found', name)
+
+            # Add objects
+            for item in mw.all_files():
+                if item.type == 'Directory' or item.use != 'original':
+                    continue
+                attrs = {
```

Comment:
The way it is is probably better. I was wondering if the properties could be given a default value of None, but it sounds like that's not workable.